rust/extractor/src/translate.rs

```diff
@@ -216,47 +217,59 @@ impl CrateTranslator<'_> {
         field_expr: &RecordLitField,
         body: &Body,
         source_map: &BodySourceMap,
-    ) -> trap::Label {
+    ) -> generated::RecordLitFieldTrapLabel {
         let RecordLitField { name, expr } = field_expr;
         let expr_label = self.emit_expr(*expr, body, source_map);
         let ret = self.trap.emit(generated::RecordLitField {
             id: TrapId::Star,
             name: name.as_str().into(),
             expr: expr_label,
         });
-        self.emit_location_for_expr(ret, *expr, source_map);
+        // TODO find a way for emitting a location for the whole thing
+        self.emit_location_for_expr(ret.into(), *expr, source_map);
         ret
     }
-    fn emit_pat(&mut self, pat_id: PatId, body: &Body, source_map: &BodySourceMap) -> trap::Label {
+    fn emit_pat(
+        &mut self,
+        pat_id: PatId,
+        body: &Body,
+        source_map: &BodySourceMap,
+    ) -> generated::PatTrapLabel {
         let pat = &body.pats[pat_id];
-        let ret = match pat {
-            ra_ap_hir_def::hir::Pat::Missing => {
-                self.trap.emit(generated::MissingPat { id: TrapId::Star })
-            }
-            ra_ap_hir_def::hir::Pat::Wild => {
-                self.trap.emit(generated::WildPat { id: TrapId::Star })
-            }
+        let ret: generated::PatTrapLabel = match pat {
+            ra_ap_hir_def::hir::Pat::Missing => self
+                .trap
+                .emit(generated::MissingPat { id: TrapId::Star })
+                .into(),
+            ra_ap_hir_def::hir::Pat::Wild => self
+                .trap
+                .emit(generated::WildPat { id: TrapId::Star })
+                .into(),
             ra_ap_hir_def::hir::Pat::Tuple { args, ellipsis } => {
                 let args = args
                     .into_iter()
                     .map(|pat| self.emit_pat(*pat, body, source_map))
                     .collect();
                 let ellipsis_index = ellipsis.and_then(|x| x.try_into().ok());
-                self.trap.emit(generated::TuplePat {
-                    id: TrapId::Star,
-                    args,
-                    ellipsis_index,
-                })
+                self.trap
```

Comment:
Can't we avoid all the `into()` calls. They are a bit annoying to write and make things less type safe. Would it be possible to have the following type for `trap.emit()` ?

```
fn emit<T>(node: T) -> Label<T>
```
That way the type checker knows that `emit(generated::TuplePat{})` returns a `Label<generated::TuplePat>`
```

Reply:
That's what's actually happening already (with `T::Label` rather than `Label<T>` because I encountered some problems with the latter). So `emit` gives the very precise type. The problem is that then we need to convert that to a more generic label, for example from a `TuplePat::Label` to a `Pat::Label`, which from the TRAP point of view is perfectly valid. As rust abhors implicit conversions, in that case an `into()` must be used.

However, that doesn't mean it's unsafe, as an `into()` from `X::Label` to `Y::Label` will only compile if `X` is a subclass of `Y`.

Reply:
I see, it feels a bit more complicated then what I would do in Haskell, but it should indeed work ;-)

Reply:
I'll still play around with it for a bit, maybe I'll get to do a more ergonomic `Label<T>`. But in any case because it's Rust we won't get rid of the `into()`s.

Reply:
Ah, I managed to get it into a much better spot, using `Label<T>` (the problem was the missing `PhantomData` and the wrong derived `Clone` and `Copy`). And, marking `from_untyped` explicitly `unsafe`, which I think works quite well to prevent accidental use :)